compiler/ilgen/VirtualMachineOperandStack.cpp

```diff
@@ -0,0 +1,197 @@
+   _mb(mb),
+   _stackTopRegister(stackTopRegister),
+   _stackMax(sizeHint),
+   _stackTop(-1),
+   _elementType(elementType)
+   {
+   int32_t numBytes = _stackMax * sizeof(TR::IlValue *);
+   _stack = (TR::IlValue **) TR::comp()->trMemory()->allocateHeapMemory(numBytes);
+   memset(_stack, 0, numBytes);
+
+   // store current operand stack pointer base address so we can use it whenever we need
+   // to recreate the stack as the interpreter would have
+   mb->Store("OperandStack_base", stackTopRegister->Load(mb));
+
+   _pushAmount = (growsUp()) ? +1 : -1;
+   _stackOffset = stackPtrStartingOffset();
+   }
+
+VirtualMachineOperandStack::VirtualMachineOperandStack(OMR::VirtualMachineOperandStack *other)
+   : VirtualMachineState(),
+   _mb(other->_mb),
+   _stackTopRegister(other->_stackTopRegister),
+   _stackMax(other->_stackMax),
+   _stackTop(other->_stackTop),
+   _elementType(other->_elementType),
+   _pushAmount(other->_pushAmount)
+   {
+   int32_t numBytes = _stackMax * sizeof(TR::IlValue *);
+   _stack = (TR::IlValue **) TR::comp()->trMemory()->allocateHeapMemory(numBytes);
+   memcpy(_stack, other->_stack, numBytes);
+   }
+
+
+// commits the simulated operand stack of values to the virtual machine state
+// the given builder object is where the operations to commit the state will be inserted
+// the top of the stack is assumed to be managed independently, most likely
+//    as a VirtualMachineRegister or a VirtualMachineRegisterInStruct
+void
+VirtualMachineOperandStack::Commit(TR::IlBuilder *b)
+   {
+   TR::IlType *Element = _elementType;
+   TR::IlType *pElement = _mb->typeDictionary()->PointerTo(Element);
+
+   TR::IlValue *stack = b->Load("OperandStack_base");
+   for (int32_t i = _stackTop;i >= 0;i--)
+      {
+      // TBD: how to handle hitting the end of stack?
+
+      b->StoreAt(
+      b->   IndexAt(pElement,
+               stack,
+      b->      ConstInt32(i - _stackOffset)),
+            Pick(i)); // should generalize, maybe delegate element storage ?
+      }
+   }
+
+void
+VirtualMachineOperandStack::MergeInto(OMR::VirtualMachineOperandStack *other, TR::IlBuilder *b)
+   {
+   TR_ASSERT(_stackTop == other->_stackTop, "stacks are not same size");
+   for (int32_t i=_stackTop;i >= 0;i--)
+      {
+      // only need to do something if the two entries aren't already the same
+      if (other->_stack[i]->getCPIndex() != _stack[i]->getCPIndex())
+         {
+         // what if types don't match? could use ConvertTo, but seems...arbitrary
+         // nobody *should* design bytecode set where corresponding elements of stacks from
+         // two incoming control flow edges can have different primitive types. objects, sure
+         // but not primitive types! Expecting to be disappointed here some day...
+         TR_ASSERT(_stack[i]->getSymbol()->getDataType() == other->_stack[i]->getSymbol()->getDataType(), "disappointing stack merge: primitive type mismatch at same depth stack elements");
```

Comment:
ah! I missed the one in the assert message. I can fix that, but perhaps I could do it as part of another PR since it's not a functional issue, just to avoid another testing go-around on this one?